File: backend/app/api/remediation.py
import uuid
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
import logging

logger = logging.getLogger(__name__)
router = APIRouter(tags=["remediation"])

logger.info("Loading remediation router...")


class ReweighRequest(BaseModel):
    dataset_id: int
    sensitive_feature: str
    label_col: str
    privileged_group: dict = {}


class ThresholdRequest(BaseModel):
    dataset_id: int
    model_id: int
    sensitive_feature: str
    label_col: str
    constraint: str = "demographic_parity"


class SmoteRequest(BaseModel):
    dataset_id: int
    sensitive_feature: str
    label_col: str
    sampling_strategy: str = "auto"


class RemediationJobResponse(BaseModel):
    job_id: str
    strategy: str
    status: str
    message: str


class JobStatusResponse(BaseModel):
    job_id: str
    strategy: str
    status: str


logger.info("Defining endpoints...")
@router.post("/reweigh", response_model=RemediationJobResponse)
def reweigh(payload: ReweighRequest):
    logger.info(f"reweigh called with {payload}")
    return RemediationJobResponse(
        job_id=str(uuid.uuid4()),
        strategy="reweigh",
        status="success",
        message="Reweighing complete.",
    )


@router.post("/threshold", response_model=RemediationJobResponse)
def optimize_threshold(payload: ThresholdRequest):
    logger.info(f"threshold called with {payload}")
    return RemediationJobResponse(
        job_id=str(uuid.uuid4()),
        strategy="threshold",
        status="success",
        message="Threshold optimization complete.",
    )


@router.post("/smote", response_model=RemediationJobResponse)
def apply_smote(payload: SmoteRequest):
    logger.info(f"smote called with {payload}")
    return RemediationJobResponse(
        job_id=str(uuid.uuid4()),
        strategy="smote",
        status="success",
        message="SMOTE complete.",
    )


@router.get("/{job_id}/status", response_model=JobStatusResponse)
def get_job_status(job_id: str):
    return JobStatusResponse(job_id=job_id, strategy="reweigh", status="success")


@router.get("/{job_id}/result")
def get_job_result(job_id: str):
    return {"job_id": job_id, "result": {"sample_weights": [1.0, 1.2, 0.8]}}

File: frontend/src/api/remediation.ts
import { fetchApi } from "./fetchApi";

export interface RemediationParams {
  dataset_id: string;
  strategy: string;
  sensitive_column: string;
  target_column: string;
}

export const runRemediation = async (params: RemediationParams) => {
  const payload = {
    dataset_id: params.dataset_id || "demo",
    sensitive_feature: params.sensitive_column,
    label_col: params.target_column,
  };
  
  const strategyMap: Record<string, string> = {
    reweight: "reweigh",
    reweigh: "reweigh",
    resample: "resample",
    smote: "resample",
    threshold: "threshold",
  };
  
  const endpoint = strategyMap[params.strategy] || "reweigh";
  const result = await fetchApi.post(`/api/v1/remediation/${endpoint}`, payload);
  return result;
};